include/openssl/hpke.h

```diff
@@ -0,0 +1,356 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/**
+ * @file
+ * Data structures/prototypes for HPKE (RFC9180)
+ */
+
+#ifndef HPKE_H_INCLUDED
+# define HPKE_H_INCLUDED
+
+# include <openssl/ssl.h>
+
+/** biggest/default buffer for keys and internal buffers we use */
+# ifndef HPKE_MAXSIZE
+#  define HPKE_MAXSIZE (2 * 1024) /* 2k: enough for anyone :-) */
+# endif
+
+/*
+ * The HPKE modes
+ */
+# define HPKE_MODE_BASE              0 /**< Base mode  */
+# define HPKE_MODE_PSK               1 /**< Pre-shared key mode */
+# define HPKE_MODE_AUTH              2 /**< Authenticated mode */
+# define HPKE_MODE_PSKAUTH           3 /**< PSK+authenticated mode */
+
+/*
+ * The (16bit) HPKE algorithn IDs
+ */
+# define HPKE_KEM_ID_RESERVED         0x0000 /**< not used */
+# define HPKE_KEM_ID_P256             0x0010 /**< NIST P-256 */
+# define HPKE_KEM_ID_P384             0x0011 /**< NIST P-256 */
+# define HPKE_KEM_ID_P521             0x0012 /**< NIST P-521 */
+# define HPKE_KEM_ID_25519            0x0020 /**< Curve25519 */
+# define HPKE_KEM_ID_448              0x0021 /**< Curve448 */
+
+# define HPKE_KDF_ID_RESERVED         0x0000 /**< not used */
+# define HPKE_KDF_ID_HKDF_SHA256      0x0001 /**< HKDF-SHA256 */
+# define HPKE_KDF_ID_HKDF_SHA384      0x0002 /**< HKDF-SHA512 */
+# define HPKE_KDF_ID_HKDF_SHA512      0x0003 /**< HKDF-SHA512 */
+# define HPKE_KDF_ID_MAX              0x0003 /**< HKDF-SHA512 */
+
+# define HPKE_AEAD_ID_RESERVED        0x0000 /**< not used */
+# define HPKE_AEAD_ID_AES_GCM_128     0x0001 /**< AES-GCM-128 */
+# define HPKE_AEAD_ID_AES_GCM_256     0x0002 /**< AES-GCM-256 */
+# define HPKE_AEAD_ID_CHACHA_POLY1305 0x0003 /**< Chacha20-Poly1305 */
+# define HPKE_AEAD_ID_MAX             0x0003 /**< Chacha20-Poly1305 */
+
+/* strings for modes */
+# define HPKE_MODESTR_BASE       "base"    /**< base mode (1), no sender auth */
+# define HPKE_MODESTR_PSK        "psk"     /**< psk mode (2) */
+# define HPKE_MODESTR_AUTH       "auth"    /**< auth (3) with sender-key pair */
+# define HPKE_MODESTR_PSKAUTH    "pskauth" /**< psk+sender-key pair (4) */
+
+/* strings for suites */
+# define HPKE_KEMSTR_P256        "p256"              /**< KEM id 0x10 */
+# define HPKE_KEMSTR_P384        "p384"              /**< KEM id 0x11 */
+# define HPKE_KEMSTR_P521        "p521"              /**< KEM id 0x12 */
+# define HPKE_KEMSTR_X25519      "x25519"            /**< KEM id 0x20 */
+# define HPKE_KEMSTR_X448        "x448"              /**< KEM id 0x21 */
+# define HPKE_KDFSTR_256         "hkdf-sha256"       /**< KDF id 1 */
+# define HPKE_KDFSTR_384         "hkdf-sha384"       /**< KDF id 2 */
+# define HPKE_KDFSTR_512         "hkdf-sha512"       /**< KDF id 3 */
+# define HPKE_AEADSTR_AES128GCM  "aes128gcm"         /**< AEAD id 1 */
+# define HPKE_AEADSTR_AES256GCM  "aes256gcm"         /**< AEAD id 2 */
+# define HPKE_AEADSTR_CP         "chachapoly1305"    /**< AEAD id 3 */
```

Comment:
There are already strings (NIDs) for all these... should those be used instead? (Or are these in the protocol itself?)

Reply:
Those strings are not in the RFC/protocol they're just for ease of use (ultimately for the command line when we get to ECH things later, but ought IMO be defined here). I tried to find good values (see another comment above) but didn't yet succeed in finding defined'd strings for 'em all. I guess this one'll need an iteration more and maybe to modify some other file(s) so good strings exist for 'em all.